esp-link/cgimega.c

```diff
@@ -0,0 +1,1384 @@
+ *
+ */
+
+#include <esp8266.h>
+#include <osapi.h>
+#include "cgi.h"
+#include "config.h"
+#include "uart.h"
+#include "stk500v2.h"
+#include "serbridge.h"
+#include "serled.h"
+
+#define INIT_DELAY     150   // wait this many millisecs before sending anything
+#define BAUD_INTERVAL  600   // interval after which we change baud rate
+#define PGM_TIMEOUT  20000   // timeout after sync is achieved, in milliseconds
+#define PGM_INTERVAL   200   // send sync at this interval in ms when in programming mode
+#define ATTEMPTS         8   // number of attempts total to make
+
+#ifdef OPTIBOOT_DBG
+#define DBG(format, ...) do { os_printf(format, ## __VA_ARGS__); } while(0)
+#else
+#define DBG(format, ...) do { } while(0)
+#endif
+
+#define DBG_GPIO5 1 // define to 1 to use GPIO5 to trigger scope
+
+//===== global state
+
+static ETSTimer optibootTimer;
+
+static enum {		// overall programming states
+  stateInit = 0,	// initial delay
+  stateSync,		// waiting to hear back
+  stateVar1,		// Steps in reading subsequent parameters
+  stateVar2,
+  stateVar3,
+  stateGetSig1,		// Reading device signature
+  stateGetSig2,
+  stateGetSig3,
+  stateGetFuse1,
+  stateGetFuse2,
+  stateGetFuse3,
+  stateProg,		// programming...
+} progState;
+static char* progStates[] = { "init", "sync", "var1", "var2", "var3",
+	"sig1", "sig2", "sig3",
+	"fuse1", "fuse2", "fuse3",
+	"prog" };
+static short baudCnt;        // counter for sync attempts at different baud rates
+static short ackWait;        // counter of expected ACKs
+static uint32_t baudRate;    // baud rate at which we're programming
+
+#define RESP_SZ 64
+static char responseBuf[RESP_SZ]; // buffer to accumulate responses from optiboot
+static short responseLen = 0;     // amount accumulated so far
+#define ERR_MAX 128
+static char errMessage[ERR_MAX];  // error message
+
+#define MAX_PAGE_SZ 512      // max flash page size supported
+#define MAX_SAVED   512      // max chars in saved buffer
+// structure used to remember request details from one callback to the next
+// allocated dynamically so we don't burn so much static RAM
+static struct optibootData {
+  char *saved;               // buffer for saved incomplete hex records
+  char *pageBuf;             // buffer for received data to be sent to AVR
+  uint16_t pageLen;          // number of bytes in pageBuf
+  uint16_t pgmSz;            // size of flash page to be programmed at a time
+  uint32_t pgmDone;          // number of bytes programmed
+  uint32_t address;          // address to write next page to
+  uint32_t segment;          // for extended segment addressing, added to the address field
+  uint32_t startTime;        // time of program POST request
+  HttpdConnData *conn;       // request doing the programming, so we can cancel it
+  bool eof;                  // got EOF record
+} *optibootData;
+
+// STK500v2 variables
+static int	hardwareVersion,
+		firmwareVersionMajor,
+		firmwareVersionMinor,
+		vTarget;
```

Comment:
Do these all need to be static (as opposed to dynamically allocated or on the stack)? They chew up RAM...